File: Leetcode/dfs_recursion_backtracking/207_Course Schedule/Solution.java
class Solution {
    public boolean canFinish(int n, int[][] prerequisites) {
        boolean[] res = {true};
        int[] visited = new int[n];
        List<ArrayList<Integer>> graph = new ArrayList<>();
        for(int i = 0; i<n; ++i){
            graph.add(new ArrayList<Integer>());
        }
        for(int[] preq: prerequisites){
            graph.get(preq[0]).add(preq[1]);
        }
        for(int i = 0; i<n; ++i){
            dfs(res, visited, i, graph);
        }
        return res[0];
    }
        private void dfs(boolean[] res, int[] visited, int src,
                         List<ArrayList<Integer>> graph){
            visited[src] = 1;
            for(int dest: graph.get(src)){
                if(visited[dest]==1){
                    res[0] = false;
                }
                if(visited[dest]==0){
                    dfs(res, visited, dest, graph);
                }
            }
            visited[src] = -1;
        }
}

File: Leetcode/dfs_recursion_backtracking/207_Course Schedule/Solution.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        n = numCourses
        graph = collections.defaultdict(list)
        for u, v in prerequisites:
            graph[u] += [v]
        res, visited = [True], [0]*n    # visit state: -1~visiting 0~unvisited 1~visited
        def dfs(u):
            visited[u] = 1
            for v in graph[u]:
                if visited[v]==1:
                    res[0] = False
                if not visited[v]:
                    dfs(v)
            visited[u] = -1
        for course in range(n):
            dfs(course)
        return res[0]
